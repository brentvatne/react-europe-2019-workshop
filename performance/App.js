import React from 'react';
import {
  Animated,
  TouchableOpacity,
  StatusBar,
  StyleSheet,
  Platform,
  Text,
  View,
} from 'react-native';
import { BorderlessButton } from 'react-native-gesture-handler';
import { Provider, connect } from 'react-redux';
import { store } from './sections/Redux';
import {
  SafeAreaView,
  createStackNavigator,
  createAppContainer,
} from 'react-navigation';
import { ScrollView, RectButton } from 'react-native-gesture-handler';
import * as Sections from './sections';
import { useScreens } from 'react-native-screens';
import { Assets as StackAssets } from 'react-navigation-stack';
import { Constants, Asset } from 'expo';
import stall from './sections/stall';

// vvvvvvv codecast url
// https://www.codecast.io/play_cast/7rGRpEMgpxy4
// ^^^^^^^ codecast url

useScreens();
Asset.loadAsync(StackAssets);

class Button extends React.Component {
  render() {
    return (
      <RectButton
        style={{
          backgroundColor: '#eee',
          borderRadius: 5,
          flex: 1,
          alignSelf: 'stretch',
          justifyContent: 'center',
          alignItems: 'center',
          marginBottom: 10,
          paddingVertical: 10,
          marginHorizontal: 20,
        }}
        onPress={this.props.onPress}>
        <Text style={{ fontSize: 15, fontWeight: '600' }}>
          {this.props.title}
        </Text>
      </RectButton>
    );
  }
}

@connect(store => ({ user: store.user }))
class Home extends React.Component {
  static navigationOptions = {
    header: null,
  };

  state = {
    options: {
      emojiSize: {
        size: 100,
      },
    },
  };

  scrollY = new Animated.Value(0);
  handleScroll = Animated.event(
    [{ nativeEvent: { contentOffset: { y: this.scrollY } } }],
    { useNativeDriver: true }
  );

  render() {
    let scale = this.scrollY.interpolate({
      inputRange: [-100, 0],
      outputRange: [1.3, 1],
      extrapolateRight: 'clamp',
    });

    let translateX = this.scrollY.interpolate({
      inputRange: [-100, 0],
      outputRange: [50, 0],
      extrapolateRight: 'clamp',
    });

    let translateY = this.scrollY.interpolate({
      inputRange: [-100, 0],
      outputRange: [-10, 0],
      extrapolateRight: 'clamp',
    });

    let opacity = this.scrollY.interpolate({
      inputRange: [0, 80],
      outputRange: [1, 0],
    });

    let underlayOpacity = this.scrollY.interpolate({
      inputRange: [0, 80],
      outputRange: [0, 1],
    });

    return (
      <SafeAreaView style={{ flex: 1 }}>
        <Animated.ScrollView
          onScroll={this.handleScroll}
          scrollEventThrottle={16}
          style={styles.container}
          contentContainerStyle={styles.contentContainer}>
          <TouchableOpacity onPress={() => stall()}>
            <Animated.Text
              style={{
                opacity,
                transform: [{ translateX }, { translateY }, { scale }],
                marginLeft: 20,
                marginBottom: 20,
                fontWeight: '600',
                fontSize: 35,
                textAlign: 'left',
              }}>
              "Performance"
            </Animated.Text>
          </TouchableOpacity>

          <Button
            title="Deferring expensive rendering"
            onPress={() => this.props.navigation.navigate('Map')}
          />

          <Button
            title="Lists"
            onPress={() => this.props.navigation.navigate('Lists')}
          />

          <Button
            title="Redux"
            onPress={() => this.props.navigation.navigate('Redux')}
          />

          <Button
            title="Animation with useNativeDriver"
            onPress={() => this.props.navigation.navigate('Animation')}
          />

          {/* {[...Array(20).keys()].map(k => (
            <Button
              key={k.toString()}
              title={`Placeholder ${k}`}
              onPress={() => this.props.navigation.navigate('Animation')}
            />
          ))} */}

          {this.props.user ? (
            <Text
              style={{
                marginTop: 15,
                color: '#888',
                textAlign: 'center',
                marginHorizontal: 20,
              }}>
              Signed in as {this.props.user.name}
            </Text>
          ) : null}
          <StatusBar hidden={false} barStyle="default" />
        </Animated.ScrollView>

        <Animated.View
          style={{
            position: 'absolute',
            top: 0,
            left: 0,
            right: 0,
            height: 25,
            backgroundColor: 'rgba(255,255,255,0.9)',
            opacity: underlayOpacity,
          }}
        />
        <BorderlessButton
          onPress={this._toggleEmojiSize}
          style={{
            position: 'absolute',
            bottom: 0,
            left: 0,
            width: 150,
            height: 150,
          }}>
          <Emoji emojiSize={this.state.options.emojiSize} />
        </BorderlessButton>
      </SafeAreaView>
    );
  }

  _toggleEmojiSize = () => {
    let options = {
      ...this.state.options,
      emojiSize: { ...this.state.options.emojiSize },
    };

    if (options.emojiSize.size === 100) {
      options.emojiSize.size = 50;
    } else {
      options.emojiSize.size = 100;
    }
    this.setState({ options });
  };
}

class Emoji extends React.PureComponent {
  render() {
    // stall(1000);

    return (
      <Text
        style={{
          fontSize: this.props.emojiSize.size,
          position: 'absolute',
          bottom: 15,
          left: 15,
        }}>
        ðŸ™ƒ
      </Text>
    );
  }
}

let Stack = createStackNavigator({
  Home,
  ...Sections,
});

let AppContainer = createAppContainer(Stack);

export default class App extends React.Component {
  render() {
    return (
      <Provider store={store}>
        <AppContainer />
      </Provider>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  contentContainer: {
    paddingTop: 50,
    paddingBottom: 30,
  },
});
