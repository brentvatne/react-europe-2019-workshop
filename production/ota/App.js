import React from 'react';
import { Constants, Updates } from 'expo';
import {
  Alert,
  Platform,
  ScrollView,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import { RectButton } from 'react-native-gesture-handler';
import SafeAreaView from 'react-native-safe-area-view';

import Sentry from 'sentry-expo';
// import { SentrySeverity, SentryLog } from 'react-native-sentry';
Sentry.config(
  'https://341188480a36496a8a983eccdecaa895@sentry.io/1460585'
).install();

export default class App extends React.Component {
  render() {
    return (
      <View style={{ flex: 1 }}>
        <ScrollView style={styles.container}>
          <SafeAreaView
            style={{
              paddingHorizontal: 15,
              paddingBottom: 100,
            }}>
            <Text style={{ fontSize: 20, marginTop: 10, fontWeight: 'bold' }}>
              Manifest for version:{' '}
              <MonoText style={{ fontSize: 16 }}>
                {Constants.manifest.version}
              </MonoText>
            </Text>
            <View
              style={{
                flex: 1,
                height: 1,
                backgroundColor: '#ccc',
                marginTop: 10,
                marginBottom: 20,
              }}
            />
            <MonoText>{JSON.stringify(Constants.manifest, null, 2)}</MonoText>
          </SafeAreaView>
        </ScrollView>

        <ReloadButton onPress={this.handlePressUpdateAsync} />
      </View>
    );
  }

  handlePressUpdateAsync = async () => {
    try {
      // Fake an error!
      throw new Error('Fake error when trying to get update!');
      const update = await Updates.checkForUpdateAsync();
      if (update.isAvailable) {
        alert('Update available! Downloading...');
        await Updates.fetchUpdateAsync();
        if (await shouldReloadAsync()) {
          Updates.reloadFromCache();
        }
      } else {
        alert('No update available');
      }
    } catch (e) {
      // Catch but don't crash!
      Sentry.captureException(e);
      alert(e.message);
    }
  };
}

const MonoText = props => (
  <Text
    {...props}
    style={[
      props.style,
      { fontFamily: Platform.select({ ios: 'Menlo', default: 'monospace' }) },
    ]}
  />
);

function shouldReloadAsync() {
  return new Promise((resolve, reject) => {
    Alert.alert(
      'Update ready',
      'Do you want to restart and open it now?',
      [
        { text: 'Yes please', onPress: () => resolve(true) },
        {
          text: 'Nope',
          onPress: () => reject(false),
          style: 'cancel',
        },
      ],
      { cancelable: false }
    );
  });
}

class ReloadButton extends React.Component {
  render() {
    return (
      <RectButton
        style={{
          position: 'absolute',
          bottom: 0,
          height: 90,
          backgroundColor: 'rgba(0,0,0,0.9)',
          right: 0,
          left: 0,
        }}>
        <RectButton
          underlayColor="rgba(255,255,255,0.7)"
          style={{
            flex: 1,
            alignItems: 'center',
            justifyContent: 'center',
          }}
          onPress={this.props.onPress}>
          <SafeAreaView>
            <Text style={{ fontSize: 18, fontWeight: '600', color: '#fff' }}>
              Check for app update
            </Text>
          </SafeAreaView>
        </RectButton>
      </RectButton>
    );
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
});
