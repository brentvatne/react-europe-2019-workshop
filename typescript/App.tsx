import React from 'react';
import { Button, StatusBar, StyleSheet, Text, View } from 'react-native';
import {
  createAppContainer,
  createStackNavigator,
  NavigationState,
  NavigationScreenProp,
} from 'react-navigation';
import { Haptic } from 'expo';

type HomeScreenProps = {
  navigation: NavigationScreenProp<NavigationState>;
};

class Home extends React.Component<HomeScreenProps> {
  static navigationOptions = {
    title: 'Home',
  };

  render() {
    return (
      <View style={styles.container}>
        <Text style={{ fontSize: 20, marginBottom: 20 }}>Home screen ðŸ˜˜</Text>
        <Button
          title="Beep boop"
          onPress={() => {
            // problem with type definition, patch-package!
            Haptic.impact(Haptic.ImpactFeedbackStyle.Heavy);
          }}
        />
        <Button
          title="Go to other"
          onPress={() =>
            this.props.navigation.navigate('Other', { message: 'ðŸ¤©' })
          }
        />
        <StatusBar barStyle="default" />
      </View>
    );
  }
}

type OtherParams = {
  message: string;
};

type OtherScreenProps = {
  navigation: NavigationScreenProp<NavigationState, OtherParams>;
};

class Other extends React.Component<OtherScreenProps> {
  static navigationOptions = {
    title: 'Other',
  };

  render() {
    return (
      <View style={styles.container}>
        <Text style={{ fontSize: 100 }}>
          {this.props.navigation.getParam('message')}
        </Text>
      </View>
    );
  }
}

let Stack = createStackNavigator(
  {
    Home,
    Other,
  },
  {
    defaultNavigationOptions: {
      title: 'Hello',
    },
  }
);

export default createAppContainer(Stack);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
