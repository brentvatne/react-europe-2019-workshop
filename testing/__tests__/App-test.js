import * as React from 'react';
import { render, fireEvent } from 'react-native-testing-library';
import App, { incrementCount } from '../App';

// // Basic unit test
test('incrementCount should increment count on state', () => {
  let initialState = { count: 5 };
  let nextState = incrementCount(initialState);
  expect(nextState.count).toEqual(6);
});

// // Snapshot test
test('should render', () => {
  expect(render(<App />).toJSON()).toMatchSnapshot();
});

// Interact with a rendered component
test('should increase count on click', () => {
  const { getByTestId, getByText } = render(<App />);
  expect(getByTestId('counter').props.children).toEqual(0);
  fireEvent.press(getByText('Increase'));
  expect(getByTestId('counter').props.children).toEqual(1);
});

// // Use mocks to test that props are invoked correctly
test('should fire onCountIncreased callback', () => {
  let onCountIncreased = jest.fn();
  const { getByTestId, getByText } = render(
    <App onCountIncreased={onCountIncreased} />
  );
  fireEvent.press(getByText('Increase'));
  fireEvent.press(getByText('Increase'));
  expect(onCountIncreased).toHaveBeenCalledTimes(2);
  expect(onCountIncreased).toHaveBeenNthCalledWith(1, 1);
  expect(onCountIncreased).toHaveBeenNthCalledWith(2, 2);
});
