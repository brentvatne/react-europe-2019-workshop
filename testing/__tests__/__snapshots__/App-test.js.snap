// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Increase
      </Text>
    </View>
  </View>
  <Text
    testID="counter"
  >
    0
  </Text>
</View>
`;
